web3/contract.py

```diff
@@ -383,20 +386,29 @@ def eventFilter(self, event_name, filter_params={}):
             argument_filter_names,
         )
 
-        data_filter_set, event_filter_params = construct_event_filter_params(
+        check_for_forbidden_api_filter_arguments(event_abi, argument_filters)
+
+        _, event_filter_params = construct_event_filter_params(
             event_abi,
             contract_address=self.address,
             argument_filters=argument_filters,
             **filter_meta_params
         )
 
+        filter_builder = EventFilterBuilder(event_abi)
+        filter_builder.address = event_filter_params.get('address')
```

Comment:
This seems like the place where it should default to the contract address if an explicit address was not provided.

Additionally, it seems like when building a filter from a contract instance we should **require** the address be the contract address and only allow the address to be specified for contract factory based filters.